### ISSUE TITLE
Get Quartets, new implementation

### ISSUE THREAD

**Issue by LinguList** [pylogeny/concordance#3]
I was confused by the edge_quarted_dict function, becuase it has lots of continues which stop the code from executing.

My suggestion is this one, using our new Tree class from `pylotree`.

```python
def quartet_dictionary(tree):
    if not isinstance(tree, Tree):
        tree = Tree(tree)    

    def as_graph(tree):
        G = {node: [] for node in tree.tree_dict}
        for name, node in tree.tree_dict.items():
            if node.ancestor:
                G[name] += [node.ancestor]
            if node.descendants:
                G[name] += node.descendants

        return G

    quartets = {}
    for name, (nodeA, nodeB, nodeC) in [(name, nodes) for name, nodes in as_graph(tree).items()
            if len(nodes) == 3]:
        print(name, nodeA, nodeB, nodeC)
        if nodeA.descendants:
            quartets[name, nodeA.name] = nodeA.descendants + [nodeB, nodeC]
        if nodeB.descendants:
            quartets[name, nodeB.name] = nodeB.descendants + [nodeA, nodeC]
        if nodeC.descendants:
            quartets[name, nodeC.name] = nodeC.descendants + [nodeA, nodeB]
    return quartets
```

**Comment by LinguList** [comment]
```python
In [12]: from pylotree import Tree

In [13]: tree = Tree("((A,B)E1,(C,D)E2,E)root;")

In [14]: quartet_dictionary(tree)
root Node("E1") Node("E2") Node("E")
E1 Node("root") Node("A") Node("B")
E2 Node("root") Node("C") Node("D")
Out[14]: 
{('root', 'E1'): [Node("A"), Node("B"), Node("E2"), Node("E")],
 ('root', 'E2'): [Node("C"), Node("D"), Node("E1"), Node("E")],
 ('E1', 'root'): [Node("E1"), Node("E2"), Node("E"), Node("A"), Node("B")],
 ('E2', 'root'): [Node("E1"), Node("E2"), Node("E"), Node("C"), Node("D")]}
```

**Comment by LinguList** [comment]
This implementation has one problem, as we still assume a rooted tree. Will try to modify.

**Comment by LinguList** [comment]
@PhyloStar, I have a new suggestion now, but we need to agree on the formats. In order to get all quartets, as in [Minh et al. (2020)](https://doi.org/10.1093/molbev/msaa106), the function to get all quartets should output all leaves as a group and split the tree in four parts there, in all possible combinations.

So I expect:

```python
get_quartets("((A,B),(C,D));")
```
to return
```python
[["A", "B", "C", "D"]]
```

**Comment by LinguList** [comment]
My implementation for this would first represent a tree as a graph:

```python
def as_graph(tree):
    """
    Return a Graph representation of a tree, modeled as dictionary.
    """
    G = {node: [] for node in tree.tree_dict}
    for name, node in tree.tree_dict.items():
        if node.ancestor and node.descendants:
            G[name] += [node.ancestor.name]
        if node.descendants:
            G[name] += [n.name for n in node.descendants]

    return G
```

**Comment by LinguList** [comment]
Then I have a function `split_by_node`, which returns all splits for a given node in a tree, and the splits are based on visiting all edges, to which the node is attached, and cutting them, returning two tree graphs:

```python
def split_by_node(node, graph):
    out = []
    nodes = graph[node]
    if len(nodes) == 1:
        return []
    splits = []
    for nodeB in nodes:
        splits += [[node, nodeB]]
    for nodeA, nodeB in splits:
        graphA, graphB = {nodeA: []}, {nodeB: []}
        queue = [(graphA, nodeA, 1, []), (graphB, nodeB, 2, [])]
        while queue:
            g, n, t, vis = queue.pop()
            for nB in graph.get(n, []):
                if t == 1 and (nB in graphB or nB == nodeA):
                    pass
                elif t == 2 and (nB in graphA or nB == nodeB):
                    pass
                else:
                    try:
                        if not nB in g[n]:
                            g[n] += [nB]
                    except KeyError:
                        g[n] = [nB]
                    if nB not in vis and graph[nB]:
                        vis += [nB]
                        queue += [(g, nB, t, vis)]
                    else:
                        g[nB] = []
        out += [(nodeA, nodeB, graphA, graphB)]
    return out
```

**Comment by LinguList** [comment]
Get quartets then iterates over all nodes and tries to split the splits one more time:

```python
def get_quartets(tree):
    
    graph = as_graph(tree)
    leaves = [node for node, links in graph.items() if not links]
    out = []
    for node in tree.tree_dict:
        for nodeA, nodeB, graphA, graphB in split_by_node(node,
                graph):
            if graphA and graphB:
                for nodeA1, nodeA2, graphA1, graphA2 in split_by_node(nodeA,
                        graphA):
                    A, B = [get_descendants(g, leaves) for g in [graphA1, graphA2]]

                    for nodeB1, nodeB2, graphB1, graphB2 in split_by_node(nodeB,
                            graphB):
                        C, D = [get_descendants(g, leaves) for g in [graphB1, graphB2]]
                        out += [[A, B, C, D]]         

    return out
```

The function `get_descendants` is a convenience function to get the leaves in the graph:

```python
def get_descendants(graph, leaves):
    """
    Return all descendants in a subgraph which are also leaves.
    """
    out = []
    for node, descendants in graph.items():
        if node in leaves:
            out += [node]
        for descendant in descendants:
            if descendant in leaves:
                out += [descendant]
    return set(out)

```

**Comment by LinguList** [comment]
But somehow this doesn't work as expected.

**Comment by LinguList** [comment]
If you have a simpler function, could you share it? The one that is currently used in `concordance` does not seem to work right, since it has these numerous continue statements and also a rather hacky "I go up the tree and then down", which I think is not the best way to proceed.

**Comment by PhyloStar** [comment]
> This implementation has one problem, as we still assume a rooted tree. Will try to modify.

Yes. Unrooted trees as in Minh et al. was pretty confusing to start with. Any edge that connects to the root is excluded in the quartet computation.

**Comment by LinguList** [comment]
I think I can soon propose a solution to compute all quartets, for unrootet trees as well.

**Comment by PhyloStar** [comment]
> If you have a simpler function, could you share it? The one that is currently used in `concordance` does not seem to work right, since it has these numerous continue statements and also a rather hacky "I go up the tree and then down", which I think is not the best way to proceed.

This is the only thing we have. :( I am looking at the code.

For unrooted trees: Add a parent (call VR, Minh et al. call it virtual root) to the root and when iterating over the edges, ignore the edge connecting the parent of the root to the root. You can see that any quartet with VR as a node should be ignored.

**Comment by LinguList** [comment]
I will push an example to this repository now, I think it works finally.

**Comment by LinguList** [comment]
Okay, I think I finally managed to do this.

**Comment by PhyloStar** [comment]
> Get quartets then iterates over all nodes and tries to split the splits one more time:
> 
> ```python
> def get_quartets(tree):
>     
>     graph = as_graph(tree)
>     leaves = [node for node, links in graph.items() if not links]
>     out = []
>     for node in tree.tree_dict:
>         for nodeA, nodeB, graphA, graphB in split_by_node(node,
>                 graph):
>             if graphA and graphB:
>                 for nodeA1, nodeA2, graphA1, graphA2 in split_by_node(nodeA,
>                         graphA):
>                     A, B = [get_descendants(g, leaves) for g in [graphA1, graphA2]]
> 
>                     for nodeB1, nodeB2, graphB1, graphB2 in split_by_node(nodeB,
>                             graphB):
>                         C, D = [get_descendants(g, leaves) for g in [graphB1, graphB2]]
>                         out += [[A, B, C, D]]         
> 
>     return out
> ```
> 
> The function `get_descendants` is a convenience function to get the leaves in the graph:
> 
> ```python
> def get_descendants(graph, leaves):
>     """
>     Return all descendants in a subgraph which are also leaves.
>     """
>     out = []
>     for node, descendants in graph.items():
>         if node in leaves:
>             out += [node]
>         for descendant in descendants:
>             if descendant in leaves:
>                 out += [descendant]
>     return set(out)
> ```

https://github.com/pylogeny/concordance/blob/main/src/pyloconcordance/gCF_Calc.py#L12

`labeler` function annotates each internal node as a string of leaf labels.

**Comment by LinguList** [comment]
Please check the util.py, where I add the function "get_quartets". 

After running the `pyhton setup.py develop`, you can check the function as this:

```python
from pyloconcordance.util import get_quartets
from pylotree import Tree
treeA = Tree("((A,B),(C,D));")
treeB = Tree("(((A,B),(C,D)),E);")
treeC = Tree("(((A,B),(C,D)),(E,F));")

print(get_quartets(treeA))
print(get_quartets(treeB))
print(get_quartets(treeC))
```

**Comment by LinguList** [comment]
Output is:

```python
[[['C'], ['A'], ['B'], ['D']]]
[[['C', 'D'], ['E'], ['A'], ['B']], [['E'], ['B', 'A'], ['C'], ['D']]]
[[['B', 'A'], ['F', 'E'], ['C'], ['D']], [['C', 'D'], ['F', 'E'], ['A'], ['B']], [['C', 'D'], ['E'], ['B', 'A'], ['F']]]
```

**Comment by LinguList** [comment]
So the output is a list of lists which all contain four lists of taxa. And these should be all possible quartets of a given tree.

**Comment by LinguList** [comment]
I assume, we may want to add this function directly to the tree package, as it may come in handy.

**Comment by PhyloStar** [comment]
> ```python
> In [12]: from pylotree import Tree
> 
> In [13]: tree = Tree("((A,B)E1,(C,D)E2,E)root;")
>

Here the root has three descendants: E1, E2 and E. The edge between Root and "E" cannot have any quartets. Those edges should be excluded. 

If there are polytomies in the tree. They need to be resolved too. We can do that using `    def resolve_polytomies(self):` function.
 
> In [14]: quartet_dictionary(tree)
> root Node("E1") Node("E2") Node("E")
> E1 Node("root") Node("A") Node("B")
> E2 Node("root") Node("C") Node("D")
> Out[14]: 
> {('root', 'E1'): [Node("A"), Node("B"), Node("E2"), Node("E")],
>  ('root', 'E2'): [Node("C"), Node("D"), Node("E1"), Node("E")],
>  ('E1', 'root'): [Node("E1"), Node("E2"), Node("E"), Node("A"), Node("B")],
>  ('E2', 'root'): [Node("E1"), Node("E2"), Node("E"), Node("C"), Node("D")]}
> ```

**Comment by LinguList** [comment]
Yes, for polytomies, my suggestion is to make a function that checks them, to have this done explicitly.

And for the case of E, I do not really understand: if you use an unrooted tree, you can derive two quartets from the tree: `AB,C,D,E` and `A,B,CD,E`. E will always be on its own, but one time, AB are grouped, and one time CD. The code does this now, since it unroots the tree, representing it as an undirected network.

Or is there a rule that quartets must have more than one leave, or that they should be derived from rooted trees?

**Comment by LinguList** [comment]
Because the paper by Minh et al. talks about unrootet trees, right?

**Comment by LinguList** [comment]
So if we do quartets for rooted trees alone, I'd then suggest we make two functions, one for rooted quartets, one for unrooted ones?

**Comment by PhyloStar** [comment]
> Yes, for polytomies, my suggestion is to make a function that checks them, to have this done explicitly.
> 
> And for the case of E, I do not really understand: if you use an unrooted tree, you can derive two quartets from the tree: `AB,C,D,E` and `A,B,CD,E`. E will always be on its own, but one time, AB are grouped, and one time CD. The code does this now, since it unroots the tree, representing it as an undirected network.
> 
> Or is there a rule that quartets must have more than one leave, or that they should be derived from rooted trees?

Yes. `Each internal branch x of T` as in the paper. The branch connecting E to the tree is an external branch and not internal. We will be looking at the branches ('R', 'AB') and ('R' and 'CD') and make quartets out of them.

**Comment by PhyloStar** [comment]
> Because the paper by Minh et al. talks about unrootet trees, right?

Yes. The paper also gives about rooted too. Then, they connect the root node to another virtual root which has only one descendant that is the root. That branch is ignored when enumerating the quartets.

**Comment by LinguList** [comment]
Okay, so we have to make this for the rooted and the unrooted version, and we need to split by BRANCH, so a tree like `((a,b),(c,d),e);` can actually not be split, right? Or how do I recognize external branches in an unrooted tree?

**Comment by PhyloStar** [comment]
If the descendant of a node is `True` accordung to `n.is_leaf` then, we ignore the branch.

**Comment by LinguList** [comment]
Okay, so leaves are ignored, only internal nodes can make up a quartet. But this would mean that the tree `((a,b),(c,d)` cannot have quartets, you would at least need 2 for each quartet, so this would be 8 taxa at least, right?

**Comment by LinguList** [comment]
Or is it, sorry but I am slow on this, that we compute all quartets, and ignore those quartets with size == 1 in terms of taxa, because these are the leaves?

**Comment by PhyloStar** [comment]
For `((a,b),(c,d),e);` The quartets would look like:

`[a], [b], [c,d], [e]`  branch connecting `(ab)` to the tree is being worked.

`[a, b], [e], [c], [d]`

**Comment by LinguList** [comment]
That's what the algorithm does by now.

**Comment by PhyloStar** [comment]
Sorry, should have worked with `newick` package from beginning. Used `dendropy` with all confusion.

Post-vaccine effect.

**Comment by LinguList** [comment]
By now, as far as I see, the order is only not the one, but this is arbitrary anyway, right?

**Comment by PhyloStar** [comment]
There is an order. 

`[a], [b], [c,d], [e]` is the same as `[b], [a], [c,d], [e]` or `[b], [a], [e], [c, d]`. nodes on the same side of the quartet can be swapped like `a` and `b`.

**Comment by LinguList** [comment]
So the output we have in the function is okay now?

**Comment by LinguList** [comment]
I mean: 
```python
print(get_quartets(treeB))
```

output:
```
[[['C', 'D'], ['E'], ['A'], ['B']], 
[['E'], ['B', 'A'], ['C'], ['D']]]
```

So we group CD one time and AB the other time.

**Comment by PhyloStar** [comment]
` return [[list(q) for q in quart] for quart in sorted(set(out))]`  on line 127 I think the sorting part here need to be removed? Does that make sense?

**Comment by PhyloStar** [comment]
Can you please check for this tree? `treeC = Tree("(((A,B),(C,D)),(E,F, G));")`

**Comment by LinguList** [comment]
Sorting can be removed, yes, it does not make sense here.

**Comment by LinguList** [comment]
```
[[['C', 'D'], ['A'], ['B'], ['F', 'G', 'E']],
 [['C', 'D'], ['B', 'A'], ['F'], ['G', 'E']],
 [['C', 'D'], ['G'], ['B', 'A'], ['F', 'E']],
 [['C', 'D'], ['E'], ['F', 'G'], ['B', 'A']],
 [['B', 'A'], ['D'], ['C'], ['F', 'G', 'E']]]
```

**Comment by LinguList** [comment]
We may want to check with larger trees.

**Comment by PhyloStar** [comment]
There is a polytomy in `(E, F, G)`. Then, the quartet:

` [['C', 'D'], ['B', 'A'], ['F'], ['G', 'E']]` makes sense since `CD` and `AB` are grouped together.

But ` [['C', 'D'], ['E'], ['F', 'G'], ['B', 'A']]` is not right since `CD` and `AB` are not grouped together?

` [['B', 'A'], ['D'], ['C'], ['F', 'G', 'E']]` `'C` and `D` should be together since they are on the same side of the internal branch separating `CD` from the tree.

**Comment by PhyloStar** [comment]
I am interpreting there is an order in the quartet.

**Comment by LinguList** [comment]
The code assumes binary trees, this is important, so it won't work with polytomies, you need to binarize them first. This is a design feature.

**Comment by LinguList** [comment]
so if you binarize it, the output seems pretty exhaustive:

```
Tree("(((A,B),(C,D)),((E,F),G));")

[[['C', 'D'], ['G'], ['B', 'A'], ['F', 'E']], 
[['C', 'D'], ['A'], ['B'], ['F', 'G', 'E']], 
[['B', 'A'], ['D'], ['C'], ['F', 'G', 'E']], 
[['G'], ['E'], ['C', 'D', 'B', 'A'], ['F']]]
```
It really lists all ways in which you can cut an unrootet tree exactly into four pieces, based on the condition that each piece must have at least one node (which can be a leave).

**Comment by LinguList** [comment]
Maybe, you can add some code that I can use in a similar way, where you add the version of quartets, but use the new tree class? Then we could compare the functions more directly.

**Comment by PhyloStar** [comment]
> The code assumes binary trees, this is important, so it won't work with polytomies, you need to binarize them first. This is a design feature.

Yes. I get it.

**Comment by PhyloStar** [comment]
> Maybe, you can add some code that I can use in a similar way, where you add the version of quartets, but use the new tree class? Then we could compare the functions more directly.

Yes. Let me work on it. I will push something in the evening. Thank you.

**Comment by LinguList** [comment]
cool!

**Comment by PhyloStar** [comment]
```
def get_quartets1(tree):
    internal_edges = []
    for n in tree.preorder:
        if n != tree.root and not n.is_leaf and n.ancestor != tree.root:
            internal_edges.append([n, n.ancestor]) 
                                

    quartets = []
    for edge in internal_edges:
        child, head = edge[0], edge[1]
        A, B = child.descendants
        A, B = A.get_leaf_names(), B.get_leaf_names()
        
        C = list(set(head.get_leaf_names() ) - set(child.get_leaf_names()))
        
        D = list(set(tree.root.get_leaf_names()) - set(head.get_leaf_names()))
        
        quartets.append([A, B, C, D])

    left, right = tree.root.descendants
    if not left.is_leaf and not right.is_leaf:
        A, B = left.descendants
        C, D = right.descendants
        
        A, B = A.get_leaf_names(), B.get_leaf_names()
        C, D = C.get_leaf_names(), D.get_leaf_names()
        
        quartets.append([A, B, C, D])
        
    return quartets
```

```
In [10]: get_quartets1(Tree("(((A,B),(C,D)),((E,F),G));"))                                                             
Out[10]: 
[[['A'], ['B'], ['C', 'D'], ['G', 'F', 'E']],
 [['C'], ['D'], ['B', 'A'], ['G', 'F', 'E']],
 [['E'], ['F'], ['G'], ['B', 'C', 'D', 'A']],
 [['A', 'B'], ['C', 'D'], ['E', 'F'], ['G']]]

In [12]: get_quartets1(Tree("(((A,B),(C,D)),E);"))                                                                     
Out[12]: [[['A'], ['B'], ['C', 'D'], ['E']], [['C'], ['D'], ['B', 'A'], ['E']]]


```

**Comment by PhyloStar** [comment]
How do I call `def resolve_polytomies(self):` function to resolve a `Tree` object? May it should be added to `pylotree`?

**Comment by PhyloStar** [comment]
Do we get similar results?

For the tree `tree = Tree("(((A,B)E1,(C,D)E3)E2,E)root;"`, I get these quartets.

[[['A'], ['B'], ['C', 'D'], ['E']], [['C'], ['D'], ['B', 'A'], ['E']]]

**Comment by LinguList** [comment]
Yes, same in my code.

**Comment by PhyloStar** [comment]
An unrooted tree from Yang. 
`tree = Tree("(((A, B), C), D, E);") `

ascii_string is fine. It shows three descendants.
```
In [15]: print(tree.root.ascii_art())                                                                                                                                                  
                        ┌─A
                ┌─Edge1─┤
        ┌─Edge2─┤       └─B
        │       └─C
──Root──┤
        ├─D
        └─E
```

Error here: 
```
---> 25     upperA, upperB = tree.root.descendants

ValueError: too many values to unpack (expected 2)
```

The root has three descendants. The problem is bcoz, unrooted tree has three children at the root itself. Without additional programming, a solution is to binarize the trees before computing the quartets. This unrooted behavior handling was the big reason I wanted rooted trees only which makes sense for linguistic purpose. Minh et al. developed their concordance factor unrooted trees but we can modify their concordance factor for rooted trees.

```
In [14]: tree.root.descendants                                                                                                                                                         
Out[14]: [Node("Edge2"), Node("D"), Node("E")]
```

**Comment by PhyloStar** [comment]
I think the root node has to be handled differently in an unrooted tree. May be you have better idea. The quartet for a tree `tree1 = Tree("(((A, B), C), (D1, D2), (E1, E2));") `

```
In [19]: tree1.root.descendants                                                                                                                                                        
Out[19]: [Node("Edge2"), Node("Edge3"), Node("Edge4")]
```

One solution is to get  A and B from descendants of Edge2 and C from Edge3, D from Edge 4. Swap Edge 2 and Edge3. Then swap Edge 2 with Edge 4 to extract  A and B.

```
In [21]: print(tree1.root.ascii_art())                                                                                                                                                 
                        ┌─A
                ┌─Edge1─┤
        ┌─Edge2─┤       └─B
        │       └─C
──Root──┤       ┌─D1
        ├─Edge3─┤
        │       └─D2
        │       ┌─E1
        └─Edge4─┤
                └─E2

```

**Comment by LinguList** [comment]
In fact, a tree with multiple nodes can also be handled, especially since the splitting process is far easier by now, the only problem is that you cannot nicely determine the concordance for an edge that links to polytomies, since obviously, this edge would not have quartets, but quintets or sixtets, etc.

**Comment by LinguList** [comment]
In an unrooted binary tree, internal nodes have always three nodes to which they link if they are internal, and one if they are external (their ancestor). So unrooting this tree where the root is a polytomy is already done when including the root, as the root will thus have three nodes to which it links. 

If you have a binary tree and unroot it, the root must be removed, since it would have only two descendants. This can also be done by pruning the graph, by merging nodes with two descendants only. 

But if you have an internal polytomy like `(((A,B)N1,((C,D)N2,(E,F)N3,(G,H)N4)N5)Root;`, there's the question how you measure branch concordance for the branch that connects N5 and N2.

**Comment by LinguList** [comment]
You may be right that we should just reconsider to measure branch concordance with respect to a rooted tree. This would mean, we sample one node in the tree, not a branch, and the branch is just the link between the node and its ancestor. Maybe it is the original idea you had (?). So to check if the character is concordant with this node / branch, you would sample the (potentially polytomic) group of all descendants of the node, and contrast them against the bi- or multi-partition of the remaining tree (need to think how that should be done: you go up the tree, take the siblings, and assign the rest to the final partition). So concordance would be if the character is identical for all sampled leaves under the node, without appearing in the other leaves sampled which are not under the node. This character would then define the branch in this sample. To me, this would make sense, since parsimony is always rooted anyway, at least in the way in which linguists think about it.

**Comment by LinguList** [comment]
But we should support the unrooted version as well, I'd say, also for testing. If you are thinking of glottolog as an evaluation or something, I can promise that in almost all cases where you want to plot characters on a glottolog tree, you will be disappointed.

**Comment by LinguList** [comment]
My solution (just pushed this to util.py) looks like this now:
```python
In [13]: from pyloconcordance.util import *

In [14]: tree = Tree('((A,B)N1,((C,D)N2,(E,F)N3,(G,H)N4)N5)Root;')

In [16]: for node in tree.preorder[1:]:
    ...:     if node.descendants:
    ...:         A, B = rooted_quartition(tree, node.name)
    ...:         print('{0:10} | {1:10} | {2:10}'.format(node.name, ' '.join([''.join(x) for x in A]), ' '.join([''.join(x) for x in B])))
    ...: 
    ...: 
    ...: 
    ...: 
N1         | A B        | CDEFGH    
N5         | CD EF GH   | AB        
N2         | C D        | EF GH AB  
N3         | E F        | CD GH AB  
N4         | G H        | CD EF AB  
```

**Comment by LinguList** [comment]
The first part is the relevant partition on which we need to check, the second partitions are only for sampling. One might argue, that for the concordance factor of a rooted tree and a node in that tree, we only need the first part.

**Comment by PhyloStar** [comment]
> But we should support the unrooted version as well, I'd say, also for testing. If you are thinking of glottolog as an evaluation or something, I can promise that in almost all cases where you want to plot characters on a glottolog tree, you will be disappointed.

This is the problem with glottolog trees. Polytomies would pop up even in consensus lexical trees. So, the solution was to use maximum clade credibility (MCC) tree where all nodes are binary even if all branches don't have greater than 50% support.

**Comment by PhyloStar** [comment]
> The first part is the relevant partition on which we need to check, the second partitions are only for sampling. One might argue, that for the concordance factor of a rooted tree and a node in that tree, we only need the first part.

Got it. I got confused with sampling. You mean we don't compute SCF for partitions starting from 2nd line.

**Comment by PhyloStar** [comment]
> You may be right that we should just reconsider to measure branch concordance with respect to a rooted tree. This would mean, we sample one node in the tree, not a branch, and the branch is just the link between the node and its ancestor. Maybe it is the original idea you had (?). So to check if the character is concordant with this node / branch, you would sample the (potentially polytomic) group of all descendants of the node, and contrast them against the bi- or multi-partition of the remaining tree (need to think how that should be done: you go up the tree, take the siblings, and assign the rest to the final partition). So concordance would be if the character is identical for all sampled leaves under the node, without appearing in the other leaves sampled which are not under the node. This character would then define the branch in this sample. To me, this would make sense, since parsimony is always rooted anyway, at least in the way in which linguists think about it.

Yes. We used the siblings method in dendropy here. Yes. same character under the node would be parsimonious. So cool.
